src/proxy/metadata.rs

```diff
@@ -0,0 +1,192 @@
+        }
+    }
+}
+
+pub struct ConnectionGuard<'a> {
+    /// _metric is just here to drop when we drop and record the connection close metric
+    _metric: MetricGuard<'a, ConnectionClose<'a>>,
+    ct: &'a ConnectionTracker,
+    tup: ConnectionTuple,
+}
+
+impl<'a> Drop for ConnectionGuard<'a> {
+    fn drop(&mut self) {
+        trace!(tuple=?self.tup, "forgetting connection");
+        let mut cm = self.ct.conns.lock().unwrap();
+        cm.remove(&self.tup);
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct ConnectionMetadata {
+    #[serde(default)]
+    identity: Option<Identity>,
+}
+
+/// METADATA_SERVER_IP provides the well-known metadata server IP.
+/// This is captured by the redirection.
+pub const METADATA_SERVER_IP: IpAddr = IpAddr::V4(Ipv4Addr::new(169, 254, 169, 111));
+
+pub async fn handle_metadata_lookup(
+    ct: &ConnectionTracker,
+    stream: TcpStream,
+    remote_addr: SocketAddr,
+) -> Result<(), Error> {
+    trace!(remote=%remote_addr, "metadata lookup");
+    if let Err(e) = crate::hyper_util::http1_server()
+        .half_close(true)
+        .header_read_timeout(Duration::from_secs(2))
+        .max_buf_size(8 * 1024)
+        .serve_connection(
+            stream,
+            service_fn(|req: hyper::Request<hyper::body::Incoming>| {
+                let ct = ct.clone();
+                async move {
+                    let res: Result<_, Infallible> =
+                        serve_request(&ct, remote_addr, req).or_else(|e| {
+                            Ok(crate::hyper_util::plaintext_response(
+                                StatusCode::UNPROCESSABLE_ENTITY,
+                                e.to_string(),
+                            ))
+                        });
+                    res
+                }
+            }),
+        )
+        .await
+    {
+        error!("Error while serving HTTP connection: {}", e);
+    }
+    Ok(())
+}
+
+fn serve_request(
+    ct: &ConnectionTracker,
+    remote: SocketAddr,
+    req: hyper::Request<hyper::body::Incoming>,
+) -> anyhow::Result<Response<Full<Bytes>>> {
+    let query = req.uri().query().ok_or(anyhow!("missing query"))?;
+
+    let params = url::form_urlencoded::parse(query.as_bytes())
+        .into_owned()
+        .collect::<HashMap<String, String>>();
+    let src = SocketAddr::new(
+        params
+            .get("srcip")
+            .ok_or(anyhow!("missing srcip"))?
+            .parse()?,
+        params
+            .get("srcport")
+            .ok_or(anyhow!("missing srcport"))?
+            .parse()?,
+    );
+    let dst = SocketAddr::new(
+        params
+            .get("dstip")
+            .ok_or(anyhow!("missing dstip"))?
+            .parse()?,
+        params
+            .get("dstport")
+            .ok_or(anyhow!("missing dstport"))?
+            .parse()?,
+    );
+    if remote.ip() != dst.ip() && remote.ip() != src.ip() {
+        anyhow::bail!("metadata server request must come from the src or dst address")
```

Comment:
Not sure this is guaranteed - for example if the connection is IPv6, pretty sure it won't be the case ( dual stack and we use a MDS v4 address ). 

Can we keep a pointer to the dest and source pod - and check its IPs ? 

Or just add a TODO and bug - we can do this in a later phase.

Reply:
Ran into this myself with `localhost` testing -- since we don't use 127.0.0.1 to connect to MDS. I put a TODO. I think we can handle it once the workload API supports dual stack (currently doesn't)